processing/src/main/java/org/apache/druid/frame/key/ClusterBy.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.key;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.java.util.common.guava.Comparators;
+import org.apache.druid.segment.ColumnInspector;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.RowSignature;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Describes a key used for sorting or partitioning.
+ *
+ * Keys have columns, and some of those columns may comprise a "bucket key". See {@link #getBucketByCount()} for
+ * details about bucket keys.
+ */
+public class ClusterBy
+{
+  private final List<SortColumn> columns;
+  private final int bucketByCount;
+
+  @JsonCreator
+  public ClusterBy(
+      @JsonProperty("columns") List<SortColumn> columns,
+      @JsonProperty("bucketByCount") int bucketByCount
+  )
+  {
+    this.columns = Preconditions.checkNotNull(columns, "columns");
+    this.bucketByCount = bucketByCount;
+
+    if (bucketByCount < 0 || bucketByCount > columns.size()) {
+      throw new IAE("Invalid bucketByCount [%d]", bucketByCount);
+    }
+  }
+
+  /**
+   * Create an empty key.
+   */
+  public static ClusterBy none()
+  {
+    return new ClusterBy(Collections.emptyList(), 0);
+  }
+
+  /**
+   * The columns that comprise this key, in order.
+   */
+  @JsonProperty
+  public List<SortColumn> getColumns()
+  {
+    return columns;
+  }
+
+  /**
+   * How many fields from {@link #getColumns()} comprise the "bucket key". Bucketing is like strict partitioning: all
```

Comment:
Presumably this is the length of the prefix of columns that comprise the bucket key?

Reply:
Indeed it is. I'll clarify it.